map/generate_chart.cpp

```diff
@@ -0,0 +1,127 @@
+#include "map/generate_chart.hpp"
+
+//#include "coding/png_memory_encoder.hpp"
+
+#include "base/assert.hpp"
+
+#include "std/fstream.hpp"
+
+#include "3party/agg/agg_conv_stroke.h"
+#include "3party/agg/agg_path_storage.h"
+#include "3party/agg/agg_pixfmt_rgba.h"
+#include "3party/agg/agg_rasterizer_scanline_aa.h"
+#include "3party/agg/agg_renderer_primitives.h"
+#include "3party/agg/agg_renderer_scanline.h"
+#include "3party/agg/agg_scanline_p.h"
+
+namespace
+{
+template <class TColor, class TOrder>
+struct TBlendAdaptor
+{
+  using order_type = TOrder;
+  using color_type = TColor;
+  using TValueType = typename color_type::value_type;
+  using TCalcType = typename color_type::calc_type;
+
+  enum EBaseScale
+  {
+    SCALE_SHIFT = color_type::base_shift,
+    SCALE_MASK = color_type::base_mask,
+  };
+
+  static AGG_INLINE void blend_pix(unsigned op, TValueType * p, unsigned cr, unsigned cg,
+                                   unsigned cb, unsigned ca, unsigned cover)
+  {
+    using TBlendTable = agg::comp_op_table_rgba<TColor, TOrder>;
+    if (p[TOrder::A])
+    {
+      TBlendTable::g_comp_op_func[op](p, (cr * ca + SCALE_MASK) >> SCALE_SHIFT, (cg * ca + SCALE_MASK) >> SCALE_SHIFT,
+                                      (cb * ca + SCALE_MASK) >> SCALE_SHIFT, ca, cover);
+    }
+    else
+    {
+      TBlendTable::g_comp_op_func[op](p, cr, cg, cb, ca, cover);
+    }
+  }
+};
+}  // namespace
+
+bool GenerateChart(size_t width, size_t height, vector<m2::PointD> const & geometry,
+                   vector<uint8_t> & frameBuffer)
+{
+  agg::rgba8 const kBackgroundColor(255, 255, 255, 255);
+  agg::rgba8 const kLineColor(0, 100, 100, 255);
+  double constexpr kLineWidth = 2;
+  uint32_t constexpr kBPP = 4;
+
+  if (width == 0 || height == 0 || geometry.empty())
+  {
+    ASSERT(false, ());
```

Comment:
I guess there's no sense to generate an empty image or generate a chart without geometry. GenerateChart(...) is called with such params it's a sign that something goes wrong.